src/main/java/org/broadinstitute/hellbender/utils/read/GoogleGenomicsReadToGATKReadAdapter.java

```diff
@@ -0,0 +1,664 @@
+            throw new IllegalArgumentException("Cannot set mate position to null");
+        }
+
+        setMatePosition(locatable.getContig(), locatable.getStart());
+    }
+
+    @Override
+    public void setFragmentLength( final int fragmentLength ) {
+        // May be negative if mate maps to lower position than read
+        genomicsRead.setFragmentLength(fragmentLength);
+    }
+
+    @Override
+    public void setMappingQuality( final int mappingQuality ) {
+        if ( mappingQuality < 0 || mappingQuality > 255 ) {
+            throw new IllegalArgumentException("mapping quality must be >= 0 and <= 255");
+        }
+
+        makeAlignmentIfNecessary();
+        genomicsRead.getAlignment().setMappingQuality(mappingQuality);
+    }
+
+    @Override
+    public void setBases( final byte[] bases ) {
+        genomicsRead.setAlignedSequence(bases != null ? StringUtil.bytesToString(bases) : null);
+    }
+
+    @Override
+    public void setBaseQualities( final byte[] baseQualities ) {
+        if ( baseQualities == null ) {
+            genomicsRead.setAlignedQuality(null);
+            return;
+        }
+
+        final List<Integer> convertedBaseQualities = new ArrayList<>(baseQualities.length);
+        for ( byte b : baseQualities ) {
+            if ( b < 0 ) {
+                throw new GATKException("Base quality score " + b + " is invalid");
+            }
+
+            convertedBaseQualities.add((int)b);
+        }
+
+        genomicsRead.setAlignedQuality(convertedBaseQualities.isEmpty() ? null : convertedBaseQualities);
+    }
+
+    @Override
+    public void setCigar( final Cigar cigar ) {
+        makeAlignmentIfNecessary();
+        genomicsRead.getAlignment().setCigar(cigar != null ? CigarConversionUtils.convertSAMCigarToCigarUnitList(cigar) : null);
+    }
+
+    @Override
+    public void setCigar( final String cigarString ) {
+        makeAlignmentIfNecessary();
+        genomicsRead.getAlignment().setCigar(cigarString != null ? CigarConversionUtils.convertSAMCigarToCigarUnitList(TextCigarCodec.decode(cigarString)) : null);
+    }
+
+    @Override
+    public void setReadGroup( final String readGroupID ) {
+        genomicsRead.setReadGroupId(readGroupID);
+    }
+
+    @Override
+    public void setNumberOfReadsInFragment( final int numberOfReads ) {
+        if ( numberOfReads < 1 ) {
+            throw new IllegalArgumentException("number of reads in fragment must be >= 1");
+        }
+
+        genomicsRead.setNumberReads(numberOfReads);
+    }
+
+    @Override
+    public void setReadNumber( final int readNumber) {
+        if ( readNumber < 1 ) {
+            throw new IllegalArgumentException("read number must be >= 1");
+        }
+
+        // Convert from 1-based to 0-based numbering
+        genomicsRead.setReadNumber(readNumber - 1);
+    }
+
+    @Override
+    public void setIsPaired( final boolean isPaired ) {
+        if ( isPaired ) {
+            genomicsRead.setNumberReads(2);
```

Comment:
Extracting constants is only good practice if there is some meaningful name for them -- things like `public static final int ONE = 1;` are an abomination.